rfcs/72-editor.md

```diff
@@ -0,0 +1,508 @@
+# Feature Name: `editor-foundation`
+
+## Summary
+
+This RFC outlines the foundation for Bevy's editor. It describes its basic
+components and how they will interact with Bevy's engine. Following this RFC
+will result in a minimum viable product for the Bevy Editor.
+
+This RFC does NOT cover the particular user interface (UI) design for the
+editor. Instead, it provides the building blocks for a proper UI design, with a
+minimal usable interface to place these components in context.
+
+## Motivation
+
+Unlike most major game engines, Bevy lacks an interactive editor. This has
+three major downsides:
+
+1. it increases the barrier to entry for potential new developers;
+2. this forces scene designers to learn to create scenes with code; and
+3. developers must rebuild their project before seeing any changes they make.
+
+While teams can mitigate the last two points by creating project specific
+tools, this does require them to divert resources building those tools that they
+could have spent building the game.
+
+Bevy's plugin-heavy design poses several challenges to making a traditional
+editor. Primarily, unlike other game engines, Bevy does limit developers access
+to the core game loop, similar instead to many game frameworks. While an
+interactive editor could be built on top of Bevy that takes away this freedom;
+doing so violates the spirit of Bevy's plugin-heavy design, thus robbing it of
+Bevy's expressive nature.
+
+**Bevy's editor must acknowledge the duel nature of Bevy as both a game engine
+and a game framework.**
+
+A secondary, smaller challenge, is that Bevy relies on the Cargo build system
+for modularity. An interactive-editor that is core to Bevy must both respect
+this design, and work with the limitations and abilities of Cargo.
+
+## User-facing explanation
+
+The easiest way to get started with Bevy's editor by downloading it with Cargo:
+
+```
+$ cargo install bevy
```

Comment:
Using `bevy` here is not possible because the `bevy` crate already exists.
```suggestion
$ cargo install bevy_cli
```
The actual binary being named `bevy` is good though :)

Reply:
Crates.io seems to let you have a binary and library crate have the same name. For example I just tested:

https://crates.io/crates/leiftestlibbinsplit

You can run `cargo install leiftestlibbinsplit` and run the executable and it runs just fine. And you can also make a new project and run `cargo add leiftestlibbinsplit`, and now you can `use leiftestlibbinsplit::*` in your project.

Am I missing something?

Reply:
Sorry for the late reply and not being more specific.
It's possible to have a library and binary in the same crate but I don't think that's an option for Bevy. It would require syncing the release of a new version of Bevy's cli to a release of Bevy as a whole. We technically could do this but I figure this will cause us trouble sooner or later.

Reply:
In any case I don't think having 2 things with the same name is good. This will create unnecessary confusion.